core/rfb.js

```diff
@@ -1517,61 +1490,209 @@ export default class RFB extends EventTargetMixin {
+                this._sock.rQskipBytes(2);                       // y-position
+                this._sock.rQskipBytes(2);                       // width
+                this._sock.rQskipBytes(2);                       // height
+                this._screen_flags = this._sock.rQshiftBytes(4); // flags
+            } else {
+                this._sock.rQskipBytes(16);
             }
+        }
 
-            if (this._timing.pixels >= (this._fb_width * this._fb_height)) {
-                if ((this._FBU.width === this._fb_width && this._FBU.height === this._fb_height) ||
-                    this._timing.fbu_rt_start > 0) {
-                    this._timing.full_fbu_total += this._timing.cur_fbu;
-                    this._timing.full_fbu_cnt++;
-                    Log.Info("Timing of full FBU, curr: " +
-                              this._timing.cur_fbu + ", total: " +
-                              this._timing.full_fbu_total + ", cnt: " +
-                              this._timing.full_fbu_cnt + ", avg: " +
-                              (this._timing.full_fbu_total / this._timing.full_fbu_cnt));
-                }
+        /*
+         * The x-position indicates the reason for the change:
+         *
+         *  0 - server resized on its own
+         *  1 - this client requested the resize
+         *  2 - another client requested the resize
+         */
 
-                if (this._timing.fbu_rt_start > 0) {
-                    const fbu_rt_diff = now - this._timing.fbu_rt_start;
-                    this._timing.fbu_rt_total += fbu_rt_diff;
-                    this._timing.fbu_rt_cnt++;
-                    Log.Info("full FBU round-trip, cur: " +
-                              fbu_rt_diff + ", total: " +
-                              this._timing.fbu_rt_total + ", cnt: " +
-                              this._timing.fbu_rt_cnt + ", avg: " +
-                              (this._timing.fbu_rt_total / this._timing.fbu_rt_cnt));
-                    this._timing.fbu_rt_start = 0;
-                }
+        // We need to handle errors when we requested the resize.
+        if (this._FBU.x === 1 && this._FBU.y !== 0) {
+            let msg = "";
+            // The y-position indicates the status code from the server
+            switch (this._FBU.y) {
+            case 1:
+                msg = "Resize is administratively prohibited";
+                break;
+            case 2:
+                msg = "Out of resources";
+                break;
+            case 3:
+                msg = "Invalid screen layout";
+                break;
+            default:
+                msg = "Unknown reason";
+                break;
             }
+            Log.Warn("Server did not accept the resize request: "
+                     + msg);
+        } else {
+            this._resize(this._FBU.width, this._FBU.height);
+        }
+
+        return true;
+    }
+
+    _handleDataRect() {
+        this._timing.last_fbu = (new Date()).getTime();
```

Comment:
let's just drop this timing code -- it's not really adding much, IMO.  We can always instrument other ways (esp using the built-in browser dev tools)

Reply:
same with the enc stats.

Reply:
It was really helpful at one point, and gave information that was difficult to get out of profilers (although they are admittedly much better now). If we are considering doing any potential performance impacting changes in the near future it might still be useful to have this. For example, WebAssembly is now supported at a level that it could support a good portion of noVNC users. Anybody inspired to attack that project?

Reply:
WebAssembly in noVNC has been on my radar for a while now, just need to find time.  Maybe this weekend...

I *think* the profilers should be good enough to get most of this now. /me shrugs

Reply:
Well, if you do start playing with WebAssembly let me know. I would be at least interested in seeing what you are working on and might be able to contribute next week to some degree. I might need something to procrastinate on. :-)

Reply:
I've started playing a bit.  The problem that I found is that since most of our heavy lifting is copying data around, and webassembly operates on a fixed chunk of memory, we kind-of need the chain of "consume from websocket" to "render to screen" to be written in the source that translates to webassembly.

Started playing around with that for fun (using Rust as the source language), but I also did some profiling (chrome's profiling tools are pretty good).

I recorded `ls -alR` on a decent-sized file tree (a checkout of https://github.com/openshift/origin), coming from a qemu vnc server (which likes to sends lots of tiny TIGHT updates).  TigerVNC takes around a minute to receive/render the frames in realtime.  noVNC takes similar in realtime playback.  In full-speed playback, it takes noVNC about 23 seconds.

Looking at a performance profile (plus some playing around), 14ish of those seconds are spent calling `ctx.putImageData` (a native browser method).  Another 2-5ish seconds are spent doing our actual computations in the TIGHT method (palette stuff, deflate, etc).  AFAICT (doing some small tests), putImageData is actually fairly performance compared to a basic copy written in webassemly (that test was pretty loosely performed, so I can go back and retest more expansively if needed), so I'm not sure how much we gain from conversion to webassembly.  If we're *super-heavily* receiving compressed data, is might help there, but I'd expect more modern servers to more aggressively send normal image data, which we can lean on the browser for heavily.

If people have recordings that seem like they're not performant, I'd love to take a poke at them, but it seems like we do pretty decently.

Reply:
@DirectXMan12 I also did some prototyping/experiments over the weekend (long flight from Japan). I'm going to keep discussing this in the existing asm.js issue (I renamed it to reflect WebAssembly): https://github.com/novnc/noVNC/issues/527